internal/exec/engine.go

```diff
@@ -383,56 +388,124 @@ func (e *Engine) renderConfig(cfg types.Config) (types.Config, error) {
+					return
+				}
+			case ParsingHTTPHeaders:
+				if cfgRef.HTTPHeaders != nil && len(cfgRef.HTTPHeaders) > 0 {
+					headers, ret.err = cfgRef.HTTPHeaders.Parse()
+					if ret.err != nil {
+						e.Logger.Warning("Fetching %s failed: %s", src, ret.err)
+						return
+					}
+				}
+
+			case Fetching:
+				if cfgRef.Compression != nil {
+					compression = *cfgRef.Compression
+				}
+				ret.cfg, ret.err = e.Fetcher.FetchToBuffer(*ret.uri, resource.FetchOptions{
+					Headers:     headers,
+					Compression: compression,
+				}, abort)
+				if ret.err != nil {
+					e.Logger.Warning("Fetching %s failed: %s", src, ret.err)
+					return
+				}
+
+			case CalculatingSHA512:
+				hash := sha512.Sum512(ret.cfg)
+				if ret.uri.Scheme != "data" {
+					e.Logger.Debug("fetched referenced config at %s with SHA512: %s", src, hex.EncodeToString(hash[:]))
+				} else {
+					// data url's might contain secrets
+					e.Logger.Debug("fetched referenced config from data url with SHA512: %s", hex.EncodeToString(hash[:]))
+				}
+			}
+		}
+	}
+	return
+}
+
 // fetchReferencedConfig fetches and parses the requested config.
 // cfgRef.Source must not be nil
 func (e *Engine) fetchReferencedConfig(cfgRef types.Resource) (types.Config, error) {
 	// this is also already checked at validation time
-	if cfgRef.Source == nil {
-		e.Logger.Crit("invalid referenced config: %v", errors.ErrSourceRequired)
-		return types.Config{}, errors.ErrSourceRequired
-	}
-	u, err := url.Parse(*cfgRef.Source)
+	sources, err := cfgRef.GetSources()
 	if err != nil {
 		return types.Config{}, err
 	}
-	var headers http.Header
-	if cfgRef.HTTPHeaders != nil && len(cfgRef.HTTPHeaders) > 0 {
-		headers, err = cfgRef.HTTPHeaders.Parse()
-		if err != nil {
-			return types.Config{}, err
-		}
-	}
-	compression := ""
-	if cfgRef.Compression != nil {
-		compression = *cfgRef.Compression
+	if len(sources) == 0 {
+		e.Logger.Crit("invalid referenced config: %v", errors.ErrSourceRequired)
+		return types.Config{}, errors.ErrSourceRequired
 	}
-	rawCfg, err := e.Fetcher.FetchToBuffer(*u, resource.FetchOptions{
-		Headers:     headers,
-		Compression: compression,
-	})
-	if err != nil {
-		return types.Config{}, err
+
+	ret := make(chan fetchResult, len(sources))
+	abort := make(chan int)
+	for _, src := range sources {
+		go func(src string) {
+			ret <- e.fetchResource(cfgRef, src, abort)
+		}(string(src))
 	}
 
-	hash := sha512.Sum512(rawCfg)
-	if u.Scheme != "data" {
-		e.Logger.Debug("fetched referenced config at %s with SHA512: %s", *cfgRef.Source, hex.EncodeToString(hash[:]))
-	} else {
-		// data url's might contain secrets
-		e.Logger.Debug("fetched referenced config from data url with SHA512: %s", hex.EncodeToString(hash[:]))
+	closed := false
+	var data fetchResult
+	for range sources {
+		d := <-ret
+		if closed {
+			continue
+		}
+		data = d
+		if data.err == nil {
+			closed = true
+			close(abort)
+		}
+	}
+	if data.err != nil {
+		return types.Config{}, data.err
```

Comment:
Hmm, I wonder if we can't reduce some of the duplicated bits into a generic function which takes a list of sources (and maybe a `Fetcher`) and have it handle setting up the channels and performing the fetches returning basically the output of the winning `FetchToBuffer` call (`[]byte, error`). Then we can just normally process the output.

In the case of this function I think we'd pre-setup the `FetchOptions`, call this generic fetch function passing in the `Fetcher` & the `FetchOptions`, and then just check the hash of the winning contents.

Reply:
yep, that's in progress